File: Go/calculator_tests.go
/*--------------------------------------------------------------------------*\
::
::  Copyright © 2023 Steffen Liersch
::  https://www.steffen-liersch.de/
::
\*--------------------------------------------------------------------------*/

package main

import "fmt"

type test struct {
	Name       string   `json:"name"`
	Expression []string `json:"expression"`
	Output     []string `json:"output"`
}

func performTests(tests []test) bool {
	testCount := 0
	errorCount := 0
	calculator := CreateCalculator(CreateFloatFormatter("%G"))
	for _, test := range tests {
		fmt.Printf("Test: %s\n", test.Name)
		for _, expr := range test.Expression {
			lines := calculator.CalculateAndFormat(expr, "")
			if !assertEqual(test.Output, lines) {
				errorCount++
			}
			testCount++
		}
	}
	fmt.Printf("=> %d tests completed with %d errors\n", testCount, errorCount)
	return errorCount <= 0
}

func assertEqual(expected []string, actual []string) bool {
	c1 := len(expected)
	c2 := len(actual)
	if c1 != c2 {
		fmt.Println("Assertion failed: Arrays with different lengths", expected, actual)
		return false
	}

	for i := 0; i < c1; i++ {
		if expected[i] != actual[i] {
			fmt.Println("Assertion failed", expected[i], actual[i])
			return false
		}
	}

	return true
}

File: CSharp/Calculator.Tests/CalculatorTests.cs
/*--------------------------------------------------------------------------*\
::
::  Copyright © 2023 Steffen Liersch
::  https://www.steffen-liersch.de/
::
\*--------------------------------------------------------------------------*/

using System;
using System.Collections.Generic;
using System.Linq;
using Liersch.Json;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Liersch.Calculator.Tests;

[TestClass]
public sealed class CalculatorTests
{
  [TestMethod]
  public void PerformTests()
  {
    int testCount = 0;
    int errorCount = 0;
    var calculator = new Calculator(new FloatFormatter("G"));
    string json = Helpers.LoadTests();
    var tests = JsonNode.Parse(json);
    foreach(JsonNode test in tests)
    {
      Console.WriteLine("Test: " + test["name"].AsString);
      IList<string> output = ToStringArray(test["output"]);
      IList<string> array = ToStringArray(test["expression"]);
      foreach(string expr in array)
      {
        IList<string> lines = calculator.CalculateAndFormat(expr, "");
        if(!AssertEqual(output, lines))
          errorCount++;
        testCount++;
      }
    }
    Console.WriteLine($"=> {testCount} tests completed with {errorCount} errors");

    if(errorCount > 0)
      Assert.Fail();
  }

  static bool AssertEqual(IList<string> expected, IList<string> actual)
  {
    int c1 = expected.Count;
    int c2 = actual.Count;
    if(c1 != c2)
    {
      Console.WriteLine("Assertion failed: Arrays with different lengths");
      return false;
    }

    for(int i = 0; i < c1; i++)
    {
      if(expected[i] != actual[i])
      {
        Console.WriteLine($"Assertion failed: {expected[i]} != {actual[i]}");
        return false;
      }
    }

    return true;
  }

  static IList<string> ToStringArray(JsonNode node)
  {
    if(node.IsArray)
      return node.Select(x => x.AsString).ToList();

    return new string[] { node.AsString };
  }
}
